crates/bevy_ecs/src/world/mod.rs

```diff
@@ -262,6 +264,23 @@ impl World {
             .init_component_with_descriptor(&mut self.storages, descriptor)
     }
 
+    /// Initializes a new dynamic component type with the given storage type.
+    ///
+    /// Every call to this method will return a new unique [`TypedComponentId`]
+    /// which allows for accessing values of this component without unsafe code.
+    pub fn init_dynamic_component<T: Send + Sync + 'static>(
+        &mut self,
+        storage_type: StorageType,
+    ) -> TypedComponentId<T> {
```

Comment:
Maybe you could have a `init_dynamic_component` that takes a `T: Component` and a `init_dynamic_component_with_storage` that only takes `T: Send + Sync + 'static` but requires specifying the storage?

Reply:
the first just sounds like init_component. Should that just return the typed wrapper then, since the inner id is still accessible?

Reply:
Changing `init_component` to return `TypedComponent<T>` might be controversial, although consistent with this change.